### ISSUE TITLE
Reader qExec could not find Shared Local Snapshot! issue on 5.0.0-alpha 6/7

### ISSUE THREAD

**Issue by amilkh** [greenplum-db/gpdb#2568]
While testing with gpdb 5.0-alpha.6 and alpha.7 (pre-release concourse build), we encountered the following issue:

```sql
drop table if exists bigtest;
create table bigtest (
	numb	integer,
	str1	varchar(20),
	str2	varchar(20),
	str3	varchar(20)
) distributed by (numb);


create or replace function fill_bigtest(num integer) returns integer as '
	declare
		count integer;
	begin
		count := 0;
		loop
			count = count + 1;
			insert into bigtest (numb, str1, str2, str3)  values (count, ''xxxxxxxxxxxxxxxxxxy'', ''asdfasdfasdfasdf'', ''asdfasdfasdfasdf'');
			exit when count >= num;
		end loop;
		return count;
	end
' language 'plpgsql';
select fill_bigtest(50000);
\c
select count(*) from bigtest A inner join bigtest B on B.str1 similar to '(x+x+)+y';

ERROR:  failed to acquire resources on one or more segments
DETAIL:  FATAL:  Reader qExec could not find Shared Local Snapshot! (sharedsnapshot.c:602)
DETAIL:  Tried to find a shared snapshot slot with id: 1408 and found none. Shared Local Snapshots dump: Local SharedSnapshot Slot Dump: currSlots: 0 maxSlots: 500
HINT:  Either this Reader qExec was created before the Writer qExec or the Writer qExec died.
 (seg0 host1:40000)
```

We haven't tested this query with alpha.4 or alpha.5 but this was working fine until alpha.3.

---
Please note that running all the queries in the same session (without the '\c' in the middle) does not throw any exception.

---
Here is the stack trace
```
2017-06-01 19:59:16.883981 EDT,"gpadmin","test",p25341,th-831469792,"host1","51250",2017-06-01 19:59:06 EDT,0,con9821,,seg0,,,,sx1,"FATAL","XX000","Reader qExec could not find Shared Local Snapshot! (sharedsnapshot.c:602)","Tried to find a shared snapshot slot with id: 9821 and found none. Shared Local Snapshots dump: Local SharedSnapshot Slot Dump: currSlots: 1 maxSlots: 500 (SLOT index: 0 slotid: 900 QDxid: 192103 QDcid: 0 pid: 4011)","Either this Reader qExec was created before the Writer qExec or the Writer qExec died.",,,,,0,,"sharedsnapshot.c",602,"Stack trace:
1    0x94bb9b postgres errstart + 0x1db
2    0x9a63ca postgres lookupSharedSnapshot (sharedsnapshot.c:595)
3    0x9630da postgres InitPostgres + 0x9fa
4    0x828598 postgres PostgresMain + 0x1a8
5    0x7c9e35 postgres <symbol not found> (postmaster.c:6364)
6    0x7cc252 postgres PostmasterMain (postmaster.c:1523)
7    0x703f97 postgres main (main.c:206)
8    0x351041ed1d libc.so.6 __libc_start_main + 0xfd
9    0x4c4f79 postgres <symbol not found> + 0x4c4f79
"
```

FYI, we have reproduced this on both SUSE11 and RHEL6 using the PivNet release of alpha 6.

Here is our setup
```sql
test=# select * from gp_segment_configuration ;
 dbid | content | role | preferred_role | mode | status | port  |    hostname     |     address
 | replication_port | san_mounts
------+---------+------+----------------+------+--------+-------+-----------------+----------------
-+------------------+------------
    1 |      -1 | p    | p              | s    | u      |  5432 | host1 | host1
 |                  |
    2 |       0 | p    | p              | s    | u      | 40000 | host1 | host1
 |                  |
(2 rows)
```

Thanks,
@amilkh @dsharp-pivotal @kaknikhil @benchristel

**Comment by amilkh** [comment]
If we disable ORCA with `ALTER DATABASE test SET OPTIMIZER = OFF;`, the query runs for over 2 hours (but we were not patient enough to see it finish and just killed it).

Also, according to @ashwinstar and @asimrp this probably won't occur if we have multiple segments.

**Comment by hsyuan** [comment]
Did you mean you run this query with Orca on, and the cluster only has 1 segment?

**Comment by xiong-gang** [comment]
I did some simple tests. Looks like we have some problem (most likely
planner bug) when the cluster has only 1 segment. Simple queries would hang.

**Comment by pengzhout** [comment]
PR https://github.com/greenplum-db/gpdb/pull/2570 is for this.

**Comment by amilkh** [comment]
Thank you all!

**Comment by d** [comment]
This is already resolved by #2570
